plugin-server/src/worker/ingestion/event-pipeline/cookielessServerHashStep.ts

```diff
@@ -0,0 +1,303 @@
+import * as crypto from 'crypto'
+import { DateTime } from 'luxon'
+// @ts-expect-error no types
+import * as siphashDouble from 'siphash/lib/siphash-double'
+import { getDomain } from 'tldts'
+
+import { ConcurrencyController } from '../../../utils/concurrencyController'
+import { DB } from '../../../utils/db/db'
+import { now } from '../../../utils/now'
+import { UUID7 } from '../../../utils/utils'
+import { EventPipelineRunner } from './runner'
+
+//---------------------------------------------------------------------
+// This pipeline step is used to get the distinct id and session id for events that are using the cookieless server hash mode.
+// At the most basic level, the new distinct id is hash(daily_salt + team_id + ip + root_domain + user_agent).
+
+// The session ID is a UUIDv7 that's stored in redis, using the timestamp of the first event of the session as the
+// timestamp part of the UUID. We implement our own session inactivity system rather than using redis TTLs, as the
+// session inactivity period is 30 minutes, and it's not impossible for us to have well over 30 minutes of ingestion lag.
+
+// The daily salt is a 128-bit random value that is stored in redis. We use the calendar day of the event in the user's
+// timezone to determine which salt to use. This means that we have more than one salt in use at any point, but once
+// a salt is expired, it is deleted, and it is impossible to recover the PII that was used to generate the hash.
+
+// Due to ingestion lag and time zones, we allow creating the hash for a calendar day that is different to the current
+// UTC calendar day, provided somewhere in the world could be in that calendar day, with some additional buffer in the
+// past for ingestion lag.
+
+// To ensure that a user that logs in and out doesn't collide with themselves, we store the set of identify event UUIDs
+// in redis against the base hash value, and use the number of identifies for that hash to calculate the final hash value.
+// The exact number we append is the number of identifies that happened before this user called identify, which will
+// mean adjusting the number based on whether the event was pre or post identify, or was itself the identify event.
+
+const TIMEZONE_FALLBACK = 'UTC'
+const SENTINEL_COOKIELESS_SERVER_HASH_DISTINCT_ID = '$sentinel_cookieless_server_hash'
+const MAX_NEGATIVE_TIMEZONE_HOURS = 12
+const MAX_POSITIVE_TIMEZONE_HOURS = 14
+const MAX_INGESTION_LAG_HOURS = 24
+const SALT_TTL_SECONDS =
+    (MAX_POSITIVE_TIMEZONE_HOURS + MAX_NEGATIVE_TIMEZONE_HOURS + MAX_INGESTION_LAG_HOURS + 24) * 60 * 60
+const SESSION_TTL_SECONDS = 60 * 60 * 24
+
+export async function cookielessServerHashStep(
+    runner: EventPipelineRunner,
+    event: PluginEvent
+): Promise<[PluginEvent | undefined]> {
+    // if events aren't using this mode, skip all processing
+    if (event.properties?.['$device_id'] !== SENTINEL_COOKIELESS_SERVER_HASH_DISTINCT_ID) {
```

Comment:
It's slightly weird that we're using `device_id` but then calling the constant `DISTINCT_ID`. Is this correct?

Reply:
You're right, this constant is potentially used in 4 places (though never all at the same time)

distinct id
anon distinct id
session id
device id

there's 3 cases:
anon event - needs to generate distinct id, device id and session id
identify - needs to generate anon distinct id, device id, and session id
post-identify - needs to generate device id, and session id

In the first case, we need *a* value for distinct id, that's originally where the idea for using this sentinel value came from. In the second case, the same is true for anon distinct id, we need *a* value, so using a sentinel value makes sense. The third case is tricker because it will have a real distinct id already, so we can't detect this mode from the distinct id / anon distinct id.

I've used device id === SENTINEL_COOKIELESS_SERVER_HASH_DISTINCT_ID to detect cookieless, because it's present on all the events.

I'd like to keep that because it's the simplest logic, but I think we could make some changes:
* send NULL as session id, just so there is 1 less thing to send
* rename the constant to something that matches how it's used

Reply:
Even though I agree with everything, it's slightly weird that we're using "sentinel value in one of these three fields" to indicate "is cookieless mode?". It'd make more sense to me if a separate value was set `'$cookieless_mode': something` instead.

When you say "need **a** value for distinct id" do you mean we need it *before* we get to this point in the codebase? Or that we *eventually* need it?

Reply:
I'm not totally clear on the details, just got a heads up from the infra team about this.

That I know of:
* distinct id is used in the kafka partitioning key, which doesn't matter here as we use a different partitioning key
* distinct id can be used to block malicious or misconfigured clients (doesn't affect us, we'd need to block by ip)
* distinct id is used in various metrics (i.e. we can see hot distinct ids, to block them)